### PR TITLE
playable youtube media atoms on fronts

### PR DESCRIPTION
**Pull request by gidsg**
## What does this change?
Adds support for playable youtube media atoms on fronts, when the *show video* option has been selected in facia-tool.

## What is the value of this and can you measure success?
Required for YouTube PFP trial feature parity.

## Does this affect other platforms - Amp, Apps, etc?
No

## Screenshots
screen shot 2017-01-09 at 17 19 33

### CODE REVIEW

File: common/app/implicits/FaciaContentFrontendHelpers.scala [hunk 1]
```diff
@@ -22,6 +24,19 @@ object FaciaContentFrontendHelpers {
       imageOverride.orElse(defaultTrailPicture)
     }
 
+    def mainYouTubeMediaAtom: Option[MediaAtom] =
+      for {
+       main <- faciaContent.properties.maybeContent.map(_.fields.main)
+       atoms <-  faciaContent.properties.maybeContent.flatMap(_.atoms)
+       document <- Some(Jsoup.parse(main))
```
Comment (gidsg) [inline common/app/implicits/FaciaContentFrontendHelpers.scala]: I am aware it's pretty ghastly to scrape HTML like this 👻  but as I couldn't manage to add the `blocks` to the content model without compilation errors to do with implicits, this is another way I can derive the information. At least it is encapsulated here so if we can get a more semantic way to work we can swap it out.  CC @TBonnin

**Comment by MatthewJWalls** [comment]
PR build results: 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-15513/screenshots/mobile.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-15513/screenshots/wide.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-15513/screenshots/tablet.png 
* https://s3-eu-west-1.amazonaws.com/prbuilds/PR-15513/screenshots/desktop.png 

 -automated message

File: static/src/stylesheets/module/facia/_container--video.scss [hunk 2]
```diff
@@ -426,3 +426,16 @@ $video-width-desktop: 700px;
...
+
+.fc-item__video-container {
+    .youtube-media-atom {
+        z-index: 1;
+    }
+
+    .youtube-media-atom__iframe:not(.youtube__video-ready) {
+        ~ .youtube-media-atom__overlay {
+            visibility: hidden;
+            opacity: 0;
```
Comment (akash1810) [inline static/src/stylesheets/module/facia/_container--video.scss]: is `opacity` needed if its hidden?

**Review by akash1810 — APPROVED** [review]
lgtm with my little knowledge on fronts 😄

**Comment by paperboyo** [comment]
Woo! Time indicator on fronts, cool! Maybe the one in carousel should be made to match design-wise?

One quick q: since these are for atoms that may not have an associated video page (?), they may also not have a separate trail image (15:9), only the video poster (16:9). Does the front automatically crop equal amounts from left and right sides of the poster image to display on 5:3 card?

Also, I remember us deleting native HTML video posters from these playables as they were unnecessarily loading for all videos, how does it work with atoms? Do they have one, separate from swanky progressively enhanced poster?

Lastly, but that's strictly frontstool arena (not sure whom to nag any more), if the atoms have no video page (hence no trail image) there would be no way to change the displayed image, because frontstool can't do a Replace image on a playable card. Just another reason to finally fix this deficiency...

**Comment by gidsg** [comment]
> One quick q: since these are for atoms that may not have an associated video page (?), they may also not have a separate trail image (15:9), only the video poster (16:9). Does the front automatically crop equal amounts from left and right sides of the poster image to display on 5:3 card?

They are links to articles with atom main media so will also have a separate trail image from the associated article.

> Woo! Time indicator on fronts, cool! Maybe the one in carousel should be made to match design-wise?

We are not updating the carousel to work with YouTube atoms for now. We may do this in future.

File: common/app/implicits/FaciaContentFrontendHelpers.scala (hunk 1 above)
Comment (gidsg) [inline common/app/implicits/FaciaContentFrontendHelpers.scala]: As discussed IRL with @TBonnin IRL, I have added this as a Todo as after some significant effort I still haven't been able to add `blocks` to the content model due to an issue deriving from the 22 fields case class limit.

File: static/src/stylesheets/module/facia/_container--video.scss (hunk 2 above)
Comment (GHaberis) [inline static/src/stylesheets/module/facia/_container--video.scss]: Opacity isn't actually needed anymore following @duarte changes in https://github.com/guardian/frontend/pull/15526. Opacity on the overlay was transitioned from 0 to 1. Now this no longer happens. I'll update this now.

Comment (duarte) [inline static/src/stylesheets/module/facia/_container--video.scss]: @GHaberis I removed the initial opacity transition from 0 to 1, the transition from 1 to 0 on play should still be there — if it doesn't make it too complex?

**Comment by GHaberis** [comment]
> One quick q: since these are for atoms that may not have an associated video page (?), they may also not have a separate trail image (15:9), only the video poster (16:9). Does the front automatically crop equal amounts from left and right sides of the poster image to display on 5:3 card?

@paperboyo yes the front does automatically crop equal amounts from left and right sides of the poster image, it's still using a background-image with background-size: cover and the position is centred (background-position: 50% 50%).

File: static/src/stylesheets/module/facia/_container--video.scss (hunk 2 above)
Comment (GHaberis) [inline static/src/stylesheets/module/facia/_container--video.scss]: @duarte no that's fine, I've tested and 1 to 0 works fine and 0 to 1 on video end too, we just don't need that initial 0.

File: common/app/implicits/FaciaContentFrontendHelpers.scala (hunk 1 above)
Comment (TBonnin) [inline common/app/implicits/FaciaContentFrontendHelpers.scala]: @mchv mention using https://github.com/xdotai/play-json-extensions to solve the play json format issue with case class with more than 22 fields

Comment (gidsg) [inline common/app/implicits/FaciaContentFrontendHelpers.scala]: OK I had a look at this library but it does not look trivial to integrate so I would rather ship this as a TODO. I have some WIP attempt at adding the `mainBlock` field in https://github.com/guardian/frontend/tree/playable-media-atoms-fronts however it has a runtime error.

**Pull request merged**